pkg/costattribution/tracker.go

```diff
@@ -0,0 +1,363 @@
+	var total float64
+	for k, v := range dict {
+		count := float64(v)
+		t.updateObservations(k, now.Unix(), 0, count, 0, nil, true)
+		total += count
+	}
+	t.updateState(now, 0, total, 0)
+}
+
+func (t *Tracker) updateCountersWithLabelAdapter(lbls []mimirpb.LabelAdapter, ts time.Time, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64, reason *string, createIfDoesNotExist bool) {
+	labelValues := t.extractLabelValuesFromLabelAdapater(lbls)
+	key := t.hashLabelValues(labelValues)
+	t.updateObservations(key, ts.Unix(), activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement, reason, createIfDoesNotExist)
+	t.updateState(ts, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement)
+}
+
+func (t *Tracker) hashLabelValues(labelValues []string) string {
+	buf := bufferPool.Get().(*bytes.Buffer)
+	buf.Reset()
+	defer bufferPool.Put(buf)
+
+	for i, value := range labelValues {
+		if i > 0 {
+			buf.WriteRune(sep)
+		}
+		buf.WriteString(value)
+	}
+	return buf.String()
+}
+
+func (t *Tracker) extractLabelValuesFromLabelAdapater(lbls []mimirpb.LabelAdapter) []string {
+	labelValues := make([]string, len(t.labels))
+	for idx, cal := range t.labels {
+		for _, l := range lbls {
+			if l.Name == cal {
+				labelValues[idx] = l.Value
+				break
+			}
+		}
+		if labelValues[idx] == "" {
+			labelValues[idx] = missingValue
+		}
+	}
+	return labelValues
+}
+
+func (t *Tracker) updateCounters(lbls labels.Labels, ts time.Time, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64, reason *string, createIfDoesNotExist bool) {
+	labelValues := make([]string, len(t.labels))
+	for idx, cal := range t.labels {
+		labelValues[idx] = lbls.Get(cal)
+		if labelValues[idx] == "" {
+			labelValues[idx] = missingValue
+		}
+	}
+	key := t.hashLabelValues(labelValues)
+	t.updateObservations(key, ts.Unix(), activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement, reason, createIfDoesNotExist)
+	t.updateState(ts, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement)
+}
+
+// updateObservations updates or creates a new observation in the 'observed' map.
+func (t *Tracker) updateObservations(key string, ts int64, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64, reason *string, createIfDoesNotExist bool) {
+	t.observedMtx.RLock()
+	o, known := t.observed[key]
+	t.observedMtx.RUnlock()
+
+	if !known {
+		if len(t.observed) < t.maxCardinality*2 && createIfDoesNotExist {
+			// When createIfDoesNotExist is false, it means that the method is called from DecrementActiveSeries, when key doesn't exist we should ignore the call
+			// Otherwise create a new observation for the key
+			t.createNewObservation(key, ts, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement, reason)
+		}
+		return
+	}
+
+	o.lastUpdate.Store(ts)
+	if activeSeriesIncrement != 0 {
+		o.activeSerie.Add(activeSeriesIncrement)
+	}
```

Comment:
We didn't check the overflow here, so we're incremeting something that isn't being used anymore, which means that the overflow number is wrong.

If we want to keep the overflow number correct, we need to handle these race conditions (and I don't think it will be easy).

Reply:
as paired, it is addressed in a2ffe5a893c9a9eef26160c760cc6e57e50a3f1d and f28d67212d212e1c43b623a47ec887b09d679f82